### DISCUSSION TITLE
[💡 Feature-Request]: Add option to remove all child nodes of a parent

### DISCUSSION THREAD

**Discussion by snoozbuster** [bcakmakoglu/vue-flow#1029] [Ideas]
Given you have two nodes such as:

```
[
  { id: 'parent' },
  { id: 'child', parentNode: 'parent' }
]
```

If you select `parent` and delete it using `Backspace` (or whatever your `deleteKeyCode` is configured to), then `child` will be deleted. However, programmatically calling `removeNodes(['parent'], true)` does _not_ remove `child`, it leaves it in the graph. In this second case, the child node has a `parentNode` property still refers to a node which now no longer exists - this also, surprisingly, does not trigger any errors or anything. 

The root cause of this is because the code responsible for hunting down child nodes and adding them to the remove set actually lives in [`Pane.vue`](https://github.com/bcakmakoglu/vue-flow/blob/master/packages/core/src/container/Pane/Pane.vue#L53), not in [`removeNodes`](https://github.com/bcakmakoglu/vue-flow/blob/master/packages/core/src/store/actions.ts#L518) as I'd have guessed. Is there some reason for these two to act differently? I would have expected that internally, `Pane` would basically just be calling `removeNodes(selectedNodes, true)` (or something very similar). Is this something that can be changed (moved into `removeNodes`)? To fix this in my application, which supports deleting nodes with either the `deleteKeyCode` or an explicit delete button, I will have to write some code which explicitly goes and cleans up child nodes, just for the explicit delete case - which not only feels messy and means more code is necessary, but is actually somewhat of a challenge in my particular case.

**Comment by bcakmakoglu** [comment]
Changing the current behavior of `removeNodes` to automatically include children would be a breaking change, so for now that won't be happening.
I might add another flag to have this feature, that would be possible - I'll have to think about that.

**Reply by snoozbuster** [reply]
I'd appreciate it - or another method `removeNodesAndChildren` or something.

**Reply by snoozbuster** [reply]
@bcakmakoglu at the very least - shouldn't this trigger an error like the following, because the child now doesn't have a parent?
![image](https://github.com/bcakmakoglu/vue-flow/assets/5778364/1a2a739b-7953-4da2-9642-824ffcbf88ab)

**Reply by bcakmakoglu** [reply]
Aren't these usually just thrown as warnings in the console? 😄 
They shouldn't interrupt any logic.

The issue with removing children, which I see atm, is what if a child that is being removed is *also* a parent of another node?
Do we loop through all of these deleting all nodes that were related/associated with the initial parent node that was deleted?

Not 100% sure on this detail tbh 🤔

**Reply by bcakmakoglu** [reply]
Try out `1.22.0` - I added a third argument to the `removeNodes` function which will *recursively* remove all child nodes of a node, meaning if a child is a parent of another node, those children will also be removed as to not leave any "orphans" on the graph.

**Reply by snoozbuster** [reply]
>Aren't these usually just thrown as warnings in the console?

I have the `@error` event hooked to Sentry/`console.error` because generally speaking I would like to avoid errors on the whole, and my assumption is that any errors that the library would emit would be issues where I am using the library incorrectly (and thus, bugs in my code that I would like to fix). 

>The issue with removing children, which I see atm, is what if a child that is being removed is also a parent of another node?
Do we loop through all of these deleting all nodes that were related/associated with the initial parent node that was deleted?

What does View.vue do if you backspace a node? I would assume that it does this, yeah. This thread is really mostly about pointing out two things:

- the inconsistency in how View.vue handles deletes with Backspace vs how removeNodes handles deletes
- the fact that `removeNodes` can produce a graph state which is invalid and would otherwise be prevented or warned about (such as if you tried to add a child node without a parent)

**Reply by bcakmakoglu** [reply]
Did you test the new arg yet and was it useful? 😄

**Reply by snoozbuster** [reply]
I haven't yet, I had some issues when trying to upgrade to the latest version and had to move on before I could resolve them.
